docs/v1/P4Runtime-Spec.mdk

```diff
@@ -1020,6 +1021,173 @@ table my_ipv4_lkup {
 }
 ~ End p4example
 
+### Structured Annotations
+
+P4 supports both unstructured and structured annotations [@P4Annotations].
+Unstructured annotations of the form `MyAnno1` or `MyAnno2(body-content)` can
+either be empty, or contain free-form content; anything between the pair of
+matched parentheses is legal. Conversely, structured annotations of the form
+`MyAnno3[]` or `MyAnno4[kvList|expressionList]` have a more prescribed syntax,
+which allows declaring key-value lists or expression lists. Both unstructured
+and structured annotations may be used simultaneously on a P4 element and
+`P4Info` supports this.
+
+The annotations described up to this point, &eg; `@brief()`, have all been
+unstructured annotations, or simply annotations. These are represented in
+`P4Info` as `repeated string annotations` fields in the various `message`s.
+Similarly, structured annotations are represented in `repeated
+StructuredAnnotation structured_annotations` fields which are siblings to the
+unstructured `annotations`. The `structured_annotations` contain parsed
+representations of the original annotation source. This parsing includes
+expression-evaluation, so the resulting `P4Info` may contain a simplified
+replica of the original structured annotations.
+
+The structured annotation messages are defined in `p4types.proto`.
+
+~ Begin Proto
+message KeyValuePair {
+  string key = 1;
+  Expression value = 2;
+}
+
+message KeyValuePairList {
+  repeated KeyValuePair kv_pairs = 1;
+}
+
+message Expression {
+  oneof value {
+    string string_value = 1;
+    int64 int64_value = 2;
+    bool bool_value = 3;
+  }
+}
+
+message ExpressionList {
+  repeated Expression expressions = 1;
+}
+
+message StructuredAnnotation {
+  string name = 1;
+  oneof body {
+    ExpressionList expression_list = 2;
+    KeyValuePairList kvpair_list = 3;
+  }
+}
+~ End Proto
+
+The `StructuredAnnotation` message can represent either a `KeyValuePairList`
+or an `ExpressionList`.
+
+The type of an expression is intentionally limited to one of the following
+base types: string literal, 64-bit signed integer, or boolean. The `p4c`
+compiler frontend which generates `P4Info` will evaluate all expressions and
+simplify them to one of the valid types. Any expressions which don't match one
+of the valid types will generate an error. For integers exceeding 64 bits,
+besides issuing an error, the compiler *could* print a suggestion to use a
+string representation, and the `P4Info` consumer could perform any necessary
+conversions.
+
+The following invariants hold:
+
+1. For any P4 entity, there are no two `StructuredAnnotation`s that have the
+same name.
+
+2. Within a `KeyValuePairList`, there are no two `KeyValuePair`s that have the
+same `key.`
+
+### Structured Annotation Examples
+
+**Empty Expression List**
+
+~ Begin P4Example
+@Empty[]
+table t {
+    ...
+}
+~ End P4Example
+
+The generated p4Info will contain:
+
+~ Begin Proto
+structured_annotations {
```

Comment:
This is missing the empty expression list.

Reply:
Good catch, thanks. Will fix.

Reply:
I personally see no reason to default to an `expression_list` here. Instead I "vote" for using an empty oneof if the structured annotation has an empty body:
```
structured_annotations {
  name: "Empty"
}
```

Reply:
I think at some point we said an empty list (since it's ambiguous) will default to an empty expression_list, but now I'm not sure why we thought that is necessary. Antonin is probably right and we should just not have the oneof.  I also don't see any text in the spec saying it defaults to an empty expression list, so that seems fine?

Reply:
@stefanheule I do not know if it is necessary, but the idea was to avoid ambiguity of whether an empty structured annotation was considered somehow both an empty expression list, and also an empty key/value pair list.  This spec pull request explicitly suggests making the empty structured annotation to be considered an empty expression list, at least partly to make the parsing grammar unambiguous: https://github.com/p4lang/p4-spec/pull/796/files

In particular, this proposed new line in that PR: "To avoid ambiguity, an empty `structuredAnnotationBody` is by definition considered to contain an empty `expressionList`."

Reply:
That said, the P4_16 language spec does not mandate how the P4Info data structure is defined.  If P4Runtime API folks would prefer a 3rd alternative in the oneof for the empty case, the P4_16 language spec writers need never be informed of that fact, and they will sleep well at night.

Reply:
Oh, I looked in the changes here to see the text for the empty structured annotation, but couldn't find it. Duh, this is the p4rt repo, not the spec, so no wonder.

As I said elsewhere, I think both are reasonable, but I strongly think they should be consistent. We are proposing a design for more structured annotations and at the same time implementing it in P4RT.  Lets make sure we know what we want, and then make it consistent in both places. This is not a case of P4-spec already having something, and the P4RT community using that in some way, we are literally designing it at the same time in both places and it's the same set of people doing the design in both places. There is no good reason in my opinion to not make the same decision in both places.

Reply:
@stefanheule good points. Someone new to this would wonder why they differed. Given the formal language reasons for disambiguating, the "safest" course is to leave the P4Info as-is with the explicit empty `ExpressionList`. This way we can be criticized for the slightly wordier corner-case of a empty StructuredAnno (but I gotta wonder - why would someone use one in the first place?), but at least we won't be criticized  for  imprecision or inconsistency.

Reply:
So in a comment in the p4c PR to support structured annotations, Mihai wrote:

> If both are empty it's the kv that is assumed to be used.

https://github.com/p4lang/p4c/pull/2227/files#r389943636

But it could be a typo, or it could be confusion on my side :)

I understand the need to disambiguate for the parser generator, I don't think it is very relevant for P4Info, but I'm fine with keeping it the way it is. I'd like us to add a sentence about this case in the P4Runtime spec though.